src/parser/core/EventOrderNormalizer.ts

```diff
@@ -0,0 +1,145 @@
+    return (
+      eo.anyTarget ||
+      (HasTarget(event1) && HasTarget(event2) && event1.targetID === event2.targetID)
+    );
+  }
+
+  normalize(events: AnyEvent[]): AnyEvent[] {
+    const fixedEvents: AnyEvent[] = [];
+
+    // loop through all events looking for re-orders
+    events.forEach((event: AnyEvent, eventIndex: number) => {
+      fixedEvents.push(event);
+
+      // check for matches of the 'before' ability
+      this.eventOrders.forEach((eo: EventOrder) => {
+        if (this._isBefore(eo, event)) {
+          // loop backwards through the event history for matches of the 'after' ability within the buffer period
+          for (
+            let previousEventIndex = eventIndex;
+            previousEventIndex >= 0;
+            previousEventIndex -= 1
+          ) {
+            const previousEvent = fixedEvents[previousEventIndex];
+            if (event.timestamp - previousEvent.timestamp > (eo.bufferMs ? eo.bufferMs : 0)) {
+              break;
+            }
+            if (
+              this._isAfter(eo, previousEvent) &&
+              this._sourceCheck(eo, event, previousEvent) &&
+              this._targetCheck(eo, event, previousEvent)
+            ) {
+              fixedEvents.splice(previousEventIndex, 1);
+              fixedEvents.push(previousEvent);
+              if (eo.updateTimestamp) {
+                previousEvent.timestamp = event.timestamp;
+              }
+              previousEvent.__modified = true;
+              break;
+            }
+          }
+        }
+      });
+    });
+
+    return fixedEvents;
+  }
+}
+
+export default EventOrderNormalizer;
+
+/**
+ * The specification of an event ordering to apply.
+ * The 'after' event will always be moved forward to be in front of the 'before' event.
+ * By default, the reordered events must have the same timestamp, source, and target.
+ */
+export type EventOrder = {
+  /** REQUIRED The ability id or ids of the event that should come before
+   * Null will match ANY event ID and should be reserved for special event types */
+  beforeEventId: null | number | number[];
+  /** REQUIRED The type or types of the event that should come before */
+  beforeEventType: EventType | EventType[];
+  /** REQUIRED The ability id or ids of the event that should come after.
+   * Null will match ANY event ID and should be reserved for special event types */
+  afterEventId: null | number | number[];
+  /** REQUIRED The type or types of the event that should come after */
+  afterEventType: EventType | EventType[];
+  /** The maximum allowed timestamp difference for a re-order to be made.
+   * Should be set to the minimum possible value to avoid false positives.
+   * Defaults to 0 ms when omitted. */
+  bufferMs?: number;
+  /** Iff true, the two events may be swapped even with different sources.
+   * In most cases this should be false, and will default to false when omitted */
+  anySource?: boolean;
+  // FIXME should we also check target instance?
```

Comment:
The RakeBleed normalizer matches targetInstance in addition to targetID... does anyone know why? Does anyone know when this is relevant?

Reply:
I figured it out, adds may have same targetID and must be discerned by different targetInstance. I added baseline targetInstance support to the normalizer.